File: p7/p746/MinCostClimbingStairs.java
package leetcode.p7.p746;

public class MinCostClimbingStairs {
    /**
     * DP
     * Time complexity: O(n)
     * Space complexity: O(1)
     */
    public int minCostClimbingStairs(int[] cost) {
        if (cost.length < 2) {
            return 0;
        }
        if (cost.length < 3) {
            return Math.min(cost[0], cost[1]);
        }
        for (int i = 2; i < cost.length; i++) {
            cost[i] = Math.min(cost[i - 1], cost[i - 2]) + cost[i];
        }
        return Math.min(cost[cost.length - 1], cost[cost.length - 2]);
    }
}

File: p7/p746/min_cost_climbing_stairs.py
from typing import List


class Solution:
    """
    Feb 15, 2025 13:13
    Time Complexity: O(n)
    Space Complexity: O(1)
    """

    def minCostClimbingStairs(self, cost: List[int]) -> int:
        dp = [0] * 3
        dp[0] = 0
        dp[1] = 0

        for i in range(2, len(cost) + 1):
            temp = min(dp[0] + cost[i - 2], dp[1] + cost[i - 1])
            dp[0] = dp[1]
            dp[1] = temp

        return dp[1]
